<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.433">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>HST.508 HW1</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="pset_1_files/libs/clipboard/clipboard.min.js"></script>
<script src="pset_1_files/libs/quarto-html/quarto.js"></script>
<script src="pset_1_files/libs/quarto-html/popper.min.js"></script>
<script src="pset_1_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="pset_1_files/libs/quarto-html/anchor.min.js"></script>
<link href="pset_1_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="pset_1_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="pset_1_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="pset_1_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="pset_1_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script>
window.MathJax = {
  tex: {packages: {'[+]': ['physics']}}
};
</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
<div class="quarto-alternate-formats"><h2>Other Formats</h2><ul><li><a href="pset_1.pdf"><i class="bi bi-file-pdf"></i>PDF</a></li></ul></div></div>
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">HST.508 HW1</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<p><strong>Name</strong>: David David</p>
<p><strong>Kerberos</strong>: davidgoh</p>
<div class="cell" data-execution_count="419">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib <span class="im">as</span> mpl</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> quantgenomics <span class="im">import</span> davidPlot</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> tqdm <span class="im">import</span> tqdm</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> multiprocessing</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.api <span class="im">as</span> sm</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.formula.api <span class="im">as</span> smf</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="problem-1-mutation-and-drift-for-the-two-allele-model-6-points" class="level1">
<h1>Problem 1 Mutation and Drift for the Two-Allele Model (6 points)</h1>
<blockquote class="blockquote">
<p>Consider the evolution of heterozygosity <span class="math inline">\(H_t\)</span> (and homozygosity <span class="math inline">\(G_t\)</span>) in a population subject to a random drift and mutations. Use the approach developed in class to consider a two-allele model, which has the same rate of mutation, μ, for alleles <span class="math inline">\(A_2 \to A_1\)</span> and <span class="math inline">\(A_1 \to A_2\)</span>.</p>
</blockquote>
<section id="problem-1-a" class="level2">
<h2 class="anchored" data-anchor-id="problem-1-a">Problem 1 a</h2>
<blockquote class="blockquote">
<p>Obtain an expression for <span class="math inline">\(G_{t+1}\)</span> as a function of <span class="math inline">\(G_t\)</span>, the mutation rate, and population size. To get this expression, consider all possible ways for getting two identical alleles in the <span class="math inline">\(t+1\)</span> generation, given that homozygosity (probability of two identical alleles) in generation <span class="math inline">\(t\)</span> was <span class="math inline">\(G_t\)</span>.</p>
</blockquote>
<p><img src="pset_1_assets/TwoAllele.png" class="img-fluid"></p>
<p><span class="math display">\[
\begin{aligned}
G_{t+1}&amp;=\frac{1}{2N}(1-\mu)^2\\&amp;+\frac{1}{2N}\mu^2\\&amp;+\left(1-\frac{1}{2N}\right)G_{t}(1-\mu)^2\\&amp;+\left(1-\frac{1}{2N}\right)G_{t}\mu^2\\&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)(1-\mu)\mu\\&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)\mu(1-\mu)\\
\end{aligned}
\]</span></p>
</section>
<section id="problem-1-b" class="level2">
<h2 class="anchored" data-anchor-id="problem-1-b">Problem 1 b</h2>
<blockquote class="blockquote">
<p>Expand the obtained expression while dropping the terms higher than the first power of or <span class="math inline">\(1/N\)</span>. Obtain an expression for the steady state heterozygosity <span class="math inline">\(H_{ss}\)</span> in the two-allele model. How does the change in heterozygosity in one generation subject to mutations and drift compare to the case due to drift only?</p>
</blockquote>
<p>We drop terms higher than the first power of <span class="math inline">\(\mu\)</span> or <span class="math inline">\(\frac{1}{2N}\)</span> since <span class="math inline">\(\mu \ll 1\)</span> and <span class="math inline">\(N \gg 1\)</span>.</p>
<ul>
<li><span class="math inline">\((1-\mu)^2 \approx 1-2\mu\)</span></li>
<li><span class="math inline">\(\mu^2 \approx 0\)</span></li>
<li><span class="math inline">\((1-\mu)\mu \approx \mu\)</span></li>
</ul>
<p><span class="math display">\[
\begin{aligned}
G_{t+1}&amp;=\frac{1}{2N}(1-2\mu)+\left(1-\frac{1}{2N}\right)G_{t}(1-2\mu)+\left(1-\frac{1}{2N}\right)(1-G_t)\mu+\left(1-\frac{1}{2N}\right)(1-G_t)\mu\\
&amp;=\frac{1}{2N}(1-2\mu)+\left(1-\frac{1}{2N}\right)G_{t}(1-2\mu)+2\left(1-\frac{1}{2N}\right)(1-G_t)\mu\\
&amp;= \left(\frac{1}{2N}-\frac{\mu}{N}\right)+\left(1-\frac{1}{2N}-2\mu + \frac{\mu}{N}\right)G_{t}+2\left(\mu-\frac{\mu}{2N}\right)(1-G_t)\\
&amp;\approx \frac{1}{2N}+\left(1-\frac{1}{2N}-2\mu \right)G_{t}+2\mu(1-G_t)\\
&amp;= \frac{1}{2N}+G_{t}-\frac{1}{2N}G_{t}-2\mu G_{t}+2\mu-2\mu G_t\\
G_{t+1}&amp;= \frac{1}{2N}+G_{t}-\frac{1}{2N}G_{t}+2\mu-4\mu G_t\\
\end{aligned}
\]</span></p>
<p><span class="math display">\[
\begin{aligned}
(1-H_{t+1})&amp;=\frac{1}{2N}+(1-H_{t})-\frac{1}{2N}(1-H_{t})+2\mu-4\mu (1-H_{t})\\
H_{t+1}&amp;=1-\frac{1}{2N}-(1-H_{t})+\frac{1}{2N}(1-H_{t})-2\mu+4\mu (1-H_{t})\\
&amp;=1-\frac{1}{2N}-1+H_t+\frac{1}{2N}-\frac{H_t}{2N}-2\mu + 4\mu-4\mu H_t\\
&amp;=H_t+ H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
\Delta H_t &amp;=  H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu
\end{aligned}
\]</span></p>
<p>At steady state, <span class="math inline">\(\Delta H_t = 0\)</span></p>
<p><span class="math display">\[
\begin{aligned}
0&amp;=  H_{ss}\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
H_{ss}&amp;=\frac{2\mu}{\frac{1}{2N}+4\mu}
\end{aligned}
\]</span></p>
<p>Comparing mutation and drift (M + D) with drift only (D):</p>
<p><span class="math display">\[\Delta H_t^\mathrm{D} =  -\frac{H_t}{2N}\]</span></p>
<p><span class="math display">\[
\begin{aligned}
\Delta H_t^\mathrm{M+D} &amp;=  H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
\Delta H_t^\mathrm{M+D}&amp;= \underbrace{-\frac{H_t}{2N}}_{\text{Drift}}\ \underbrace{\vphantom{-\frac{H_t}{2N}}-4\mu H_t+2\mu}_{\text{Mutation}}\\
\end{aligned}
\]</span></p>
<p>The reversible mutation adds a <span class="math inline">\(-4\mu H_t+2\mu\)</span> term to the change in heterozygosity in one generation.</p>
</section>
<section id="problem-1-c" class="level2">
<h2 class="anchored" data-anchor-id="problem-1-c">Problem 1 C</h2>
<blockquote class="blockquote">
<p>Extra credit [+2]: Now consider a k-allele model with the same rate for all possible mutations <span class="math inline">\(A_i \to A_j\)</span>. How does the number of alleles affect the steady state heterozygosity? Consider limits of <span class="math inline">\(k=2\)</span> and <span class="math inline">\(k \to \infty\)</span>.</p>
</blockquote>
<p><img src="pset_1_assets/KAllele.png" class="img-fluid"></p>
<p><span class="math display">\[
\begin{aligned}
G_{t+1} &amp;= \frac{1}{2N}(1-\mu)^2\\&amp;+\frac{1}{2N}\mu^2(k-1)\\&amp;+\left(1-\frac{1}{2N}\right)G_t(1-\mu)^2\\&amp;+\left(1-\frac{1}{2N}\right)G_t\mu^2(k-1)\\&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)(1-\mu)\mu\\&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)\mu(1-\mu)\\&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)\mu^2 (k-2)
\end{aligned}
\]</span></p>
<p>We drop terms higher than the first power of <span class="math inline">\(\mu\)</span> or <span class="math inline">\(\frac{1}{2N}\)</span> since <span class="math inline">\(\mu \ll 1\)</span> and <span class="math inline">\(N \gg 1\)</span>.</p>
<ul>
<li><span class="math inline">\((1-\mu)^2 \approx 1-2\mu\)</span></li>
<li><span class="math inline">\(\mu^2 \approx 0\)</span></li>
<li><span class="math inline">\((1-\mu)\mu \approx \mu\)</span></li>
</ul>
<p><span class="math display">\[
\begin{aligned}
G_{t+1} &amp;= \frac{1}{2N}(1-2\mu)\\
&amp;+\left(1-\frac{1}{2N}\right)G_t(1-2\mu)\\
&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)\mu\\
&amp;+\left(1-\frac{1}{2N}\right)(1-G_t)\mu
\end{aligned}
\]</span></p>
</section>
</section>
<section id="problem-2-simulations-of-genetic-drift-8-points" class="level1">
<h1>Problem 2 Simulations of Genetic Drift (8 points)</h1>
<blockquote class="blockquote">
<p>Start with a population of <span class="math inline">\(N=100\)</span> diploid individuals, each of which has two homologous chromosomes, giving you a total of <span class="math inline">\(2N\)</span> chromosomes to track. Each chromosome has a single polymorphic locus (SNP) that can be in one of two states. To code this, create an array of integers, of length <span class="math inline">\(2N\)</span>, and set the value of each element to either 1 or 2 depending on the allele of this chromosome. In the initial population, alleles 1 and 2 are equally abundant, i.e.&nbsp;<span class="math inline">\(p=q=0.5\)</span>. Please take a look at the included starter code if you need guidance (<code>pset1_sample_code.py</code>).</p>
</blockquote>
</section>
<section id="problem-2-a" class="level1">
<h1>Problem 2 a</h1>
<blockquote class="blockquote">
<p>Simulate drift in one population. Generate the next generation by drawing individuals at random from the current generation with replacement. Compute heterozygosity for each generation. Continue until all individuals become of one type – this is called fixation. Show a plot of heterozygosity <span class="math inline">\((p)\)</span> over time for 1 trajectory.</p>
</blockquote>
<div class="cell" data-execution_count="508">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> freq(alleles):</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>    n_alleles <span class="op">=</span> <span class="bu">len</span>(alleles)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>    p <span class="op">=</span> <span class="bu">sum</span>(alleles <span class="op">==</span> <span class="dv">1</span>)<span class="op">/</span>n_alleles</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> p</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> heterozygosity(alleles):</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    n_alleles <span class="op">=</span> <span class="bu">len</span>(alleles)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    p <span class="op">=</span> <span class="bu">sum</span>(alleles <span class="op">==</span> <span class="dv">1</span>)<span class="op">/</span>n_alleles</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="dv">2</span><span class="op">*</span>p<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> drift(n_individuals,p0):</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>    p0 <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>    n_chromosomes <span class="op">=</span> <span class="dv">2</span><span class="op">*</span>n_individuals</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>    alleles <span class="op">=</span> np.random.choice([<span class="dv">1</span>,<span class="dv">2</span>],</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>                            size<span class="op">=</span>n_chromosomes,</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a>                            p<span class="op">=</span>[p0,<span class="dv">1</span><span class="op">-</span>p0])</span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>    p <span class="op">=</span> freq(alleles)</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>    p_lst <span class="op">=</span> [p]</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    Ht_lst <span class="op">=</span> [<span class="dv">2</span><span class="op">*</span>p<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p)]</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> p_lst[<span class="op">-</span><span class="dv">1</span>] <span class="op">!=</span> <span class="dv">0</span> <span class="kw">and</span> p_lst[<span class="op">-</span><span class="dv">1</span>] <span class="op">!=</span> <span class="dv">1</span>:</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>        alleles <span class="op">=</span> np.random.choice(alleles,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>                        size<span class="op">=</span>n_chromosomes)</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>        p <span class="op">=</span> freq(alleles)</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>        p_lst.append(p)</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>        Ht_lst.append(<span class="dv">2</span><span class="op">*</span>p<span class="op">*</span>(<span class="dv">1</span><span class="op">-</span>p))</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> p_lst, Ht_lst</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="512">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>p_lst, _ <span class="op">=</span> drift(<span class="dv">100</span>,<span class="fl">0.5</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>plt.plot(p_lst)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Frequency of allele $A_1$, $p$"</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Generations"</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">1</span>,linestyle<span class="op">=</span><span class="st">"--"</span>, color<span class="op">=</span><span class="st">"gray"</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">0</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, color<span class="op">=</span><span class="st">"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="512">
<pre><code>&lt;matplotlib.lines.Line2D at 0x298951d50&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-4-output-2.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="problem-2-b" class="level1">
<h1>Problem 2 b</h1>
<blockquote class="blockquote">
<p>Simulate 1000 trajectories, keeping a record of the fixation time for each simulation. Compute the mean time to fixation and compare it to <span class="math inline">\(N\)</span>. You can run your simulations at 3-5 different values of <span class="math inline">\(N\)</span> to establish this dependence better.</p>
</blockquote>
<p><span class="math display">\[H_{t}=H_0\left(1-\frac{1}{2N}\right)^{t}\approx H_0 \exp\left[{-\frac{t}{2N}}\right]\]</span></p>
<div class="cell" data-execution_count="654">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>t_dct <span class="op">=</span> {}</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> N <span class="kw">in</span> [<span class="dv">100</span>,<span class="dv">500</span>,<span class="dv">1000</span>,<span class="dv">5000</span>,<span class="dv">10000</span>]:</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>    t_fixation_or_loss <span class="op">=</span> []</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> _ <span class="kw">in</span> tqdm(<span class="bu">range</span>(<span class="dv">10000</span>)):</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>        p_lst,_ <span class="op">=</span> drift(N,<span class="fl">0.5</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>        t_fixation_or_loss.append(<span class="bu">len</span>(p_lst))</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>    arr <span class="op">=</span> np.array(t_fixation_or_loss)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>    np.savetxt(<span class="ss">f"./20230916_time_to_fixation/</span><span class="sc">{</span>N<span class="sc">}</span><span class="ss">"</span>,arr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 10000/10000 [01:17&lt;00:00, 128.60it/s]
100%|██████████| 10000/10000 [16:43&lt;00:00,  9.97it/s]
  1%|          | 108/10000 [00:40&lt;1:10:54,  2.33it/s]</code></pre>
</div>
</div>
<div class="cell" data-execution_count="652">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> N, lst <span class="kw">in</span> t_dct.items():</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>    sns.histplot(lst, kde<span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-6-output-1.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="641">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>t_fixation_or_loss[<span class="op">-</span><span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="641">
<pre><code>(1000, 2)</code></pre>
</div>
</div>
</section>
<section id="problem-2-c" class="level1">
<h1>Problem 2 c</h1>
<blockquote class="blockquote">
<p>Plot 10 individual trajectories of heterozygosity vs.&nbsp;time on a single graph. In addition, for each time point, compute the average heterozygosity by averaging over all 1000 trajectories and plot the average heterozygosity vs time. To better see the exponential decay of heterozygosity you can plot the log of heterozygosity vs time. Measure the rate of decay of heterozygosity (the slope on the log-linear plot). Compare it to N.</p>
</blockquote>
<div class="cell" data-execution_count="516">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> _ <span class="kw">in</span> tqdm(<span class="bu">range</span>(<span class="dv">10</span>)):</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>    p_lst, _ <span class="op">=</span> drift(<span class="dv">100</span>,<span class="fl">0.5</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>    ax.plot(p_lst)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Frequency of allele $A_1$, $p$"</span>)</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Generations"</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">1</span>,linestyle<span class="op">=</span><span class="st">"--"</span>, color<span class="op">=</span><span class="st">"gray"</span>)</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>ax.axhline(<span class="dv">0</span>, linestyle<span class="op">=</span><span class="st">"--"</span>, color<span class="op">=</span><span class="st">"gray"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 10/10 [00:00&lt;00:00, 120.73it/s]</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="516">
<pre><code>&lt;matplotlib.lines.Line2D at 0x29b4293d0&gt;</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-8-output-3.png" class="img-fluid"></p>
</div>
</div>
<div class="cell" data-execution_count="635">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>lst <span class="op">=</span> []</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> _ <span class="kw">in</span> tqdm(<span class="bu">range</span>(<span class="dv">1000</span>)):</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>    p_lst, Ht_lst <span class="op">=</span> drift(<span class="dv">100</span>,<span class="fl">0.5</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>    lst.append(Ht_lst)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>lens <span class="op">=</span> np.array([<span class="bu">len</span>(i) <span class="cf">for</span> i <span class="kw">in</span> lst])</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Mask of valid places in each row</span></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>mask <span class="op">=</span> np.arange(lens.<span class="bu">max</span>()) <span class="op">&lt;</span> lens[:,<span class="va">None</span>]</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true" tabindex="-1"></a>out <span class="op">=</span> np.zeros(mask.shape)</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true" tabindex="-1"></a>out[mask] <span class="op">=</span> np.concatenate(lst)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true" tabindex="-1"></a>mean_heterozygosity <span class="op">=</span> out.mean(axis<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true" tabindex="-1"></a>log_mean_heterozygosity <span class="op">=</span> np.log(mean_heterozygosity)</span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true" tabindex="-1"></a>ax.plot(mean_heterozygosity)</span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="st">"Heterozygosity $H_t$"</span>)</span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Generations"</span>)</span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb13-23"><a href="#cb13-23" aria-hidden="true" tabindex="-1"></a>ax.plot(log_mean_heterozygosity)</span>
<span id="cb13-24"><a href="#cb13-24" aria-hidden="true" tabindex="-1"></a>ax.set_ylabel(<span class="vs">r"Log heterozygosity $\log\left[H_t\right]$"</span>)</span>
<span id="cb13-25"><a href="#cb13-25" aria-hidden="true" tabindex="-1"></a>ax.set_xlabel(<span class="st">"Generations"</span>)</span>
<span id="cb13-26"><a href="#cb13-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-27"><a href="#cb13-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-28"><a href="#cb13-28" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pd.DataFrame()</span>
<span id="cb13-29"><a href="#cb13-29" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"Log"</span>] <span class="op">=</span> log_mean_heterozygosity</span>
<span id="cb13-30"><a href="#cb13-30" aria-hidden="true" tabindex="-1"></a>df[<span class="st">"Generations"</span>] <span class="op">=</span> np.arange(<span class="dv">1</span>,<span class="bu">len</span>(mean_heterozygosity)<span class="op">+</span><span class="dv">1</span>)</span>
<span id="cb13-31"><a href="#cb13-31" aria-hidden="true" tabindex="-1"></a>model <span class="op">=</span> smf.ols(<span class="st">"Log ~ Generations"</span>, data <span class="op">=</span> df[:<span class="op">-</span><span class="dv">250</span>])</span>
<span id="cb13-32"><a href="#cb13-32" aria-hidden="true" tabindex="-1"></a>results <span class="op">=</span> model.fit()</span>
<span id="cb13-33"><a href="#cb13-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-34"><a href="#cb13-34" aria-hidden="true" tabindex="-1"></a>ax.plot(results.predict(df[<span class="st">"Generations"</span>]))</span>
<span id="cb13-35"><a href="#cb13-35" aria-hidden="true" tabindex="-1"></a>ax.text(<span class="dv">1000</span>,<span class="op">-</span><span class="dv">2</span>,<span class="st">"$R^2$=</span><span class="sc">{:.2}</span><span class="st">"</span>.<span class="bu">format</span>(results.rsquared))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>100%|██████████| 1000/1000 [00:07&lt;00:00, 127.22it/s]
/var/folders/_c/3843qt453ds6klt3y95rpjcw0000gn/T/ipykernel_40568/380761567.py:15: RuntimeWarning: divide by zero encountered in log
  log_mean_heterozygosity = np.log(mean_heterozygosity)</code></pre>
</div>
<div class="cell-output cell-output-display" data-execution_count="635">
<pre><code>Text(1000, -2, '$R^2$=0.99')</code></pre>
</div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-9-output-3.png" class="img-fluid"></p>
</div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-9-output-4.png" class="img-fluid"></p>
</div>
</div>
<p>The log heterozygosity is largely linear with negative gradient and implies an exponential decay of the heterozygosity. Let the rate of decay of heterozygosity be <span class="math inline">\(B\)</span>. <span class="math display">\[H_{t}= A\exp{\left(Bt\right)}\]</span> <span class="math display">\[\log H_{t}= A + Bt\]</span></p>
<p>From the formula derived in class for random mating, fixed population sizes, no mutation, no selection, we have: <span class="math display">\[H_{t}=H_0\left(1-\frac{1}{2N}\right)^{t}\approx H_0 \exp\left[{-\frac{t}{2N}}\right]\]</span></p>
<p><span class="math display">\[\log H_{t}=\log H_0 -\frac{1}{2N}t\]</span></p>
<p>Hence, we can get the population size from the rate of decay of heterozygosity <span class="math inline">\(B\)</span>: <span class="math display">\[B=-\frac{1}{2N} \implies N = -\frac{1}{2B}\]</span></p>
<div class="cell" data-execution_count="638">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>B <span class="op">=</span> results.params[<span class="st">"Generations"</span>]</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(B)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>N <span class="op">=</span> <span class="op">-</span> <span class="dv">1</span><span class="op">/</span>(<span class="dv">2</span><span class="op">*</span>B)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(N)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>93.1803651678084</code></pre>
</div>
</div>
<p>The estimated population size from linear regression of the log heterozygosity is in the correct order of magnitude:</p>
<p><span class="math display">\[N_{\mathrm{reg}} \approx 100\]</span></p>
</section>
<section id="problem-2-d" class="level1">
<h1>Problem 2 d</h1>
<blockquote class="blockquote">
<p>Consider an effect of changing population size on the time to fixation. Run simulations first at some <span class="math inline">\(N=N_1\)</span> and then change <span class="math inline">\(N\)</span> to <span class="math inline">\(N=N_2\)</span> for <span class="math inline">\(t_2\)</span> number of generations then back to <span class="math inline">\(N_1\)</span>. Try <span class="math inline">\(N_2\ll N_1\)</span>, e.g.&nbsp;<span class="math inline">\(N_2=N_1/10\)</span>. Make 1000 runs and compute the mean time to fixation. Try several <span class="math inline">\(N_1\)</span> and <span class="math inline">\(N_2\)</span>. How does decrease in the population size change the mean time to fixation? Compute <span class="math inline">\(N_\mathrm{eff}\)</span> for each parameters (<span class="math inline">\(N_1\)</span>, <span class="math inline">\(N_2\)</span>, <span class="math inline">\(t_2\)</span>), and correlated mean time to fixation with <span class="math inline">\(N_1\)</span>, <span class="math inline">\(N_2\)</span>, and <span class="math inline">\(N_\mathrm{eff}\)</span>. Interpret your findings.</p>
</blockquote>
<div class="cell" data-execution_count="574">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> driftVaryingPopulation(N0,p0,ratio,onset,t_2):</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>    p0 <span class="op">=</span> <span class="fl">0.5</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>    n_chromosomes <span class="op">=</span> <span class="dv">2</span><span class="op">*</span>N0</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>    alleles <span class="op">=</span> np.random.choice([<span class="dv">1</span>,<span class="dv">2</span>],</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>                            size<span class="op">=</span>n_chromosomes,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                            p<span class="op">=</span>[p0,<span class="dv">1</span><span class="op">-</span>p0])</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>    p_lst <span class="op">=</span> [freq(alleles)]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a>    Ht_lst <span class="op">=</span> [heterozygosity(alleles)]</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true" tabindex="-1"></a>    generation <span class="op">=</span> <span class="dv">1</span></span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true" tabindex="-1"></a>    <span class="cf">while</span> p_lst[<span class="op">-</span><span class="dv">1</span>] <span class="op">!=</span> <span class="dv">0</span> <span class="kw">and</span> p_lst[<span class="op">-</span><span class="dv">1</span>] <span class="op">!=</span> <span class="dv">1</span>:</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true" tabindex="-1"></a>        generation <span class="op">+=</span> <span class="dv">1</span></span>
<span id="cb18-12"><a href="#cb18-12" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> generation <span class="op">&lt;=</span> onset <span class="kw">or</span> generation <span class="op">&gt;=</span> onset<span class="op">+</span>t_2:</span>
<span id="cb18-13"><a href="#cb18-13" aria-hidden="true" tabindex="-1"></a>            alleles <span class="op">=</span> np.random.choice(alleles,</span>
<span id="cb18-14"><a href="#cb18-14" aria-hidden="true" tabindex="-1"></a>                            size<span class="op">=</span>n_chromosomes)</span>
<span id="cb18-15"><a href="#cb18-15" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb18-16"><a href="#cb18-16" aria-hidden="true" tabindex="-1"></a>            alleles <span class="op">=</span> np.random.choice(alleles,</span>
<span id="cb18-17"><a href="#cb18-17" aria-hidden="true" tabindex="-1"></a>                            size<span class="op">=</span><span class="bu">round</span>(n_chromosomes<span class="op">/</span>ratio))</span>
<span id="cb18-18"><a href="#cb18-18" aria-hidden="true" tabindex="-1"></a>        p_lst.append(freq(alleles))</span>
<span id="cb18-19"><a href="#cb18-19" aria-hidden="true" tabindex="-1"></a>        Ht_lst.append(heterozygosity(alleles))</span>
<span id="cb18-20"><a href="#cb18-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> p_lst, Ht_lst</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-execution_count="624">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>fig, ax <span class="op">=</span> plt.subplots()</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>p_lst, _ <span class="op">=</span> driftVaryingPopulation(<span class="dv">100</span>,<span class="fl">0.5</span>,<span class="dv">10</span>,<span class="dv">30</span>,<span class="dv">5</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>plt.plot(p_lst)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<p><img src="pset_1_files/figure-html/cell-12-output-1.png" class="img-fluid"></p>
</div>
</div>
</section>
<section id="problem-2-e" class="level1">
<h1>Problem 2 e</h1>
<blockquote class="blockquote">
<p>[Extra credit: +2 points] Plot the distribution of the fixation time for each <span class="math inline">\(N\)</span>. How broad is the distribution? Think of various ways to quantify this. Study the mean time to fixation as a function of the number K of alleles (types of individuals), starting with <span class="math inline">\(1/K\)</span> initial frequency.</p>
</blockquote>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>