% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
\PassOptionsToPackage{dvipsnames,svgnames,x11names}{xcolor}
%
\documentclass[
  letterpaper,
  DIV=11,
  numbers=noendperiod]{scrartcl}

\usepackage{amsmath,amssymb}
\usepackage{iftex}
\ifPDFTeX
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
\usepackage{lmodern}
\ifPDFTeX\else  
    % xetex/luatex font selection
  \setmainfont[]{Times New Roman}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\usepackage[top=20mm,left=20mm,bottom=20mm,right=20mm]{geometry}
\setlength{\emergencystretch}{3em} % prevent overfull lines
\setcounter{secnumdepth}{-\maxdimen} % remove section numbering
% Make \paragraph and \subparagraph free-standing
\ifx\paragraph\undefined\else
  \let\oldparagraph\paragraph
  \renewcommand{\paragraph}[1]{\oldparagraph{#1}\mbox{}}
\fi
\ifx\subparagraph\undefined\else
  \let\oldsubparagraph\subparagraph
  \renewcommand{\subparagraph}[1]{\oldsubparagraph{#1}\mbox{}}
\fi

\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{241,243,245}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.40,0.45,0.13}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\BuiltInTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\ExtensionTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.28,0.35,0.67}{#1}}
\newcommand{\ImportTok}[1]{\textcolor[rgb]{0.00,0.46,0.62}{#1}}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\NormalTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.68,0.00,0.00}{#1}}
\newcommand{\RegionMarkerTok}[1]{\textcolor[rgb]{0.00,0.23,0.31}{#1}}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.07,0.07,0.07}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.13,0.47,0.30}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.37,0.37,0.37}{\textit{#1}}}

\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}\usepackage{longtable,booktabs,array}
\usepackage{calc} % for calculating minipage widths
% Correct order of tables after \paragraph or \subparagraph
\usepackage{etoolbox}
\makeatletter
\patchcmd\longtable{\par}{\if@noskipsec\mbox{}\fi\par}{}{}
\makeatother
% Allow footnotes in longtable head/foot
\IfFileExists{footnotehyper.sty}{\usepackage{footnotehyper}}{\usepackage{footnote}}
\makesavenoteenv{longtable}
\usepackage{graphicx}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother

\usepackage{mhchem} \usepackage{physics} \usepackage{cancel} \usepackage{framed} \usepackage{xcolor} \let\oldquote=\quote \let\endoldquote=\endquote \colorlet{shadecolor}{orange!15} \renewenvironment{quote}{\begin{shaded*}\begin{oldquote}}{\end{oldquote}\end{shaded*}}
\KOMAoption{captions}{tableheading}
\makeatletter
\makeatother
\makeatletter
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\AtBeginDocument{%
\ifdefined\contentsname
  \renewcommand*\contentsname{Table of contents}
\else
  \newcommand\contentsname{Table of contents}
\fi
\ifdefined\listfigurename
  \renewcommand*\listfigurename{List of Figures}
\else
  \newcommand\listfigurename{List of Figures}
\fi
\ifdefined\listtablename
  \renewcommand*\listtablename{List of Tables}
\else
  \newcommand\listtablename{List of Tables}
\fi
\ifdefined\figurename
  \renewcommand*\figurename{Figure}
\else
  \newcommand\figurename{Figure}
\fi
\ifdefined\tablename
  \renewcommand*\tablename{Table}
\else
  \newcommand\tablename{Table}
\fi
}
\@ifpackageloaded{float}{}{\usepackage{float}}
\floatstyle{ruled}
\@ifundefined{c@chapter}{\newfloat{codelisting}{h}{lop}}{\newfloat{codelisting}{h}{lop}[chapter]}
\floatname{codelisting}{Listing}
\newcommand*\listoflistings{\listof{codelisting}{List of Listings}}
\makeatother
\makeatletter
\@ifpackageloaded{caption}{}{\usepackage{caption}}
\@ifpackageloaded{subcaption}{}{\usepackage{subcaption}}
\makeatother
\makeatletter
\@ifpackageloaded{tcolorbox}{}{\usepackage[skins,breakable]{tcolorbox}}
\makeatother
\makeatletter
\@ifundefined{shadecolor}{\definecolor{shadecolor}{rgb}{.97, .97, .97}}
\makeatother
\makeatletter
\makeatother
\makeatletter
\makeatother
\ifLuaTeX
  \usepackage{selnolig}  % disable illegal ligatures
\fi
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\urlstyle{same} % disable monospaced font for URLs
\hypersetup{
  pdftitle={HST.508 HW1},
  colorlinks=true,
  linkcolor={blue},
  filecolor={Maroon},
  citecolor={Blue},
  urlcolor={Blue},
  pdfcreator={LaTeX via pandoc}}

\title{HST.508 HW1}
\author{}
\date{}

\begin{document}
\maketitle
\ifdefined\Shaded\renewenvironment{Shaded}{\begin{tcolorbox}[enhanced, breakable, frame hidden, borderline west={3pt}{0pt}{shadecolor}, boxrule=0pt, interior hidden, sharp corners]}{\end{tcolorbox}}\fi

\textbf{Name}: David David

\textbf{Kerberos}: davidgoh

\begin{Shaded}
\begin{Highlighting}[]
\ImportTok{import}\NormalTok{ numpy }\ImportTok{as}\NormalTok{ np}
\ImportTok{import}\NormalTok{ matplotlib }\ImportTok{as}\NormalTok{ mpl}
\ImportTok{import}\NormalTok{ matplotlib.pyplot }\ImportTok{as}\NormalTok{ plt}
\ImportTok{from}\NormalTok{ quantgenomics }\ImportTok{import}\NormalTok{ davidPlot}
\ImportTok{from}\NormalTok{ tqdm }\ImportTok{import}\NormalTok{ tqdm}
\ImportTok{import}\NormalTok{ seaborn }\ImportTok{as}\NormalTok{ sns}
\ImportTok{import}\NormalTok{ multiprocessing}
\ImportTok{import}\NormalTok{ statsmodels.api }\ImportTok{as}\NormalTok{ sm}
\ImportTok{import}\NormalTok{ statsmodels.formula.api }\ImportTok{as}\NormalTok{ smf}
\ImportTok{import}\NormalTok{ pandas }\ImportTok{as}\NormalTok{ pd}
\end{Highlighting}
\end{Shaded}

\hypertarget{problem-1-mutation-and-drift-for-the-two-allele-model-6-points}{%
\section{Problem 1 Mutation and Drift for the Two-Allele Model (6
points)}\label{problem-1-mutation-and-drift-for-the-two-allele-model-6-points}}

\begin{quote}
Consider the evolution of heterozygosity \(H_t\) (and homozygosity
\(G_t\)) in a population subject to a random drift and mutations. Use
the approach developed in class to consider a two-allele model, which
has the same rate of mutation, μ, for alleles \(A_2 \to A_1\) and
\(A_1 \to A_2\).
\end{quote}

\hypertarget{problem-1-a}{%
\subsection{Problem 1 a}\label{problem-1-a}}

\begin{quote}
Obtain an expression for \(G_{t+1}\) as a function of \(G_t\), the
mutation rate, and population size. To get this expression, consider all
possible ways for getting two identical alleles in the \(t+1\)
generation, given that homozygosity (probability of two identical
alleles) in generation \(t\) was \(G_t\).
\end{quote}

\includegraphics{pset_1_1.jpg}

\[
\begin{aligned}
G_{t+1}&=\frac{1}{2N}(1-\mu)^2+\frac{1}{2N}\mu^2+\left(1-\frac{1}{2N}\right)G_{t}(1-\mu)^2+\left(1-\frac{1}{2N}\right)G_{t}\mu^2\\
&+\left(1-\frac{1}{2N}\right)(1-G_t)(1-\mu)\mu+\left(1-\frac{1}{2N}\right)(1-G_t)\mu(1-\mu)\\
\end{aligned}
\]

\hypertarget{problem-1-b}{%
\subsection{Problem 1 b}\label{problem-1-b}}

\begin{quote}
Expand the obtained expression while dropping the terms higher than the
first power of or \(1/N\). Obtain an expression for the steady state
heterozygosity \(H_{ss}\) in the two-allele model. How does the change
in heterozygosity in one generation subject to mutations and drift
compare to the case due to drift only?
\end{quote}

We drop terms higher than the first power of \(\mu\) or \(\frac{1}{2N}\)
since \(\mu \ll 1\) and \(N \gg 1\).

\begin{itemize}
\tightlist
\item
  \((1-\mu)^2 \approx 1-2\mu\)
\item
  \(\mu^2 \approx 0\)
\item
  \((1-\mu)\mu \approx \mu\)
\end{itemize}

\[
\begin{aligned}
G_{t+1}&=\frac{1}{2N}(1-2\mu)+\left(1-\frac{1}{2N}\right)G_{t}(1-2\mu)+\left(1-\frac{1}{2N}\right)(1-G_t)\mu+\left(1-\frac{1}{2N}\right)(1-G_t)\mu\\
&=\frac{1}{2N}(1-2\mu)+\left(1-\frac{1}{2N}\right)G_{t}(1-2\mu)+2\left(1-\frac{1}{2N}\right)(1-G_t)\mu\\
&= \left(\frac{1}{2N}-\frac{\mu}{N}\right)+\left(1-\frac{1}{2N}-2\mu + \frac{\mu}{N}\right)G_{t}+2\left(\mu-\frac{\mu}{2N}\right)(1-G_t)\\
&\approx \frac{1}{2N}+\left(1-\frac{1}{2N}-2\mu \right)G_{t}+2\mu(1-G_t)\\
&= \frac{1}{2N}+G_{t}-\frac{1}{2N}G_{t}-2\mu G_{t}+2\mu-2\mu G_t\\
G_{t+1}&= \frac{1}{2N}+G_{t}-\frac{1}{2N}G_{t}+2\mu-4\mu G_t\\
\end{aligned}
\]

\[
\begin{aligned}
(1-H_{t+1})&=\frac{1}{2N}+(1-H_{t})-\frac{1}{2N}(1-H_{t})+2\mu-4\mu (1-H_{t})\\
H_{t+1}&=1-\frac{1}{2N}-(1-H_{t})+\frac{1}{2N}(1-H_{t})-2\mu+4\mu (1-H_{t})\\
&=1-\frac{1}{2N}-1+H_t+\frac{1}{2N}-\frac{H_t}{2N}-2\mu + 4\mu-4\mu H_t\\
&=H_t+ H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
\Delta H_t &=  H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu
\end{aligned}
\]

At steady state, \(\Delta H_t = 0\)

\[
\begin{aligned}
0&=  H_{ss}\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
H_{ss}&=\frac{2\mu}{\frac{1}{2N}+4\mu}
\end{aligned}
\]

Comparing mutation and drift (M + D) with drift only (D):

\[\Delta H_t^\mathrm{D} =  -\frac{H_t}{2N}\]

\[
\begin{aligned}
\Delta H_t^\mathrm{M+D} &=  H_t\left(-\frac{1}{2N}-4\mu\right)+2\mu\\
\Delta H_t^\mathrm{M+D}&= \underbrace{-\frac{H_t}{2N}}_{\text{Drift}}\ \underbrace{\vphantom{-\frac{H_t}{2N}}-4\mu H_t+2\mu}_{\text{Mutation}}\\
\end{aligned}
\]

The reversible mutation adds a \(-4\mu H_t+2\mu\) term to the change in
heterozygosity in one generation.

\hypertarget{problem-1-c}{%
\subsection{Problem 1 C}\label{problem-1-c}}

\begin{quote}
Extra credit {[}+2{]}: Now consider a k-allele model with the same rate
for all possible mutations \(A_i \to A_j\). How does the number of
alleles affect the steady state heterozygosity? Consider limits of
\(k=2\) and \(k \to \infty\).
\end{quote}

\hypertarget{problem-2-simulations-of-genetic-drift-8-points}{%
\section{Problem 2 Simulations of Genetic Drift (8
points)}\label{problem-2-simulations-of-genetic-drift-8-points}}

\begin{quote}
Start with a population of \(N=100\) diploid individuals, each of which
has two homologous chromosomes, giving you a total of \(2N\) chromosomes
to track. Each chromosome has a single polymorphic locus (SNP) that can
be in one of two states. To code this, create an array of integers, of
length \(2N\), and set the value of each element to either 1 or 2
depending on the allele of this chromosome. In the initial population,
alleles 1 and 2 are equally abundant, i.e.~\(p=q=0.5\). Please take a
look at the included starter code if you need guidance
(\texttt{pset1\_sample\_code.py}).
\end{quote}

\hypertarget{problem-2-a}{%
\section{Problem 2 a}\label{problem-2-a}}

\begin{quote}
Simulate drift in one population. Generate the next generation by
drawing individuals at random from the current generation with
replacement. Compute heterozygosity for each generation. Continue until
all individuals become of one type -- this is called fixation. Show a
plot of heterozygosity \((p)\) over time for 1 trajectory.
\end{quote}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{def}\NormalTok{ freq(alleles):}
\NormalTok{    n\_alleles }\OperatorTok{=} \BuiltInTok{len}\NormalTok{(alleles)}
\NormalTok{    p }\OperatorTok{=} \BuiltInTok{sum}\NormalTok{(alleles }\OperatorTok{==} \DecValTok{1}\NormalTok{)}\OperatorTok{/}\NormalTok{n\_alleles}
    \ControlFlowTok{return}\NormalTok{ p}

\KeywordTok{def}\NormalTok{ heterozygosity(alleles):}
\NormalTok{    n\_alleles }\OperatorTok{=} \BuiltInTok{len}\NormalTok{(alleles)}
\NormalTok{    p }\OperatorTok{=} \BuiltInTok{sum}\NormalTok{(alleles }\OperatorTok{==} \DecValTok{1}\NormalTok{)}\OperatorTok{/}\NormalTok{n\_alleles}
    \ControlFlowTok{return} \DecValTok{2}\OperatorTok{*}\NormalTok{p}\OperatorTok{*}\NormalTok{(}\DecValTok{1}\OperatorTok{{-}}\NormalTok{p)}

\KeywordTok{def}\NormalTok{ drift(n\_individuals,p0):}
\NormalTok{    p0 }\OperatorTok{=} \FloatTok{0.5}
\NormalTok{    n\_chromosomes }\OperatorTok{=} \DecValTok{2}\OperatorTok{*}\NormalTok{n\_individuals}
\NormalTok{    alleles }\OperatorTok{=}\NormalTok{ np.random.choice([}\DecValTok{1}\NormalTok{,}\DecValTok{2}\NormalTok{],}
\NormalTok{                            size}\OperatorTok{=}\NormalTok{n\_chromosomes,}
\NormalTok{                            p}\OperatorTok{=}\NormalTok{[p0,}\DecValTok{1}\OperatorTok{{-}}\NormalTok{p0])}
\NormalTok{    p\_lst }\OperatorTok{=}\NormalTok{ [freq(alleles)]}
\NormalTok{    Ht\_lst }\OperatorTok{=}\NormalTok{ [heterozygosity(alleles)]}
    \ControlFlowTok{while}\NormalTok{ p\_lst[}\OperatorTok{{-}}\DecValTok{1}\NormalTok{] }\OperatorTok{!=} \DecValTok{0} \KeywordTok{and}\NormalTok{ p\_lst[}\OperatorTok{{-}}\DecValTok{1}\NormalTok{] }\OperatorTok{!=} \DecValTok{1}\NormalTok{:}
\NormalTok{        alleles }\OperatorTok{=}\NormalTok{ np.random.choice(alleles,}
\NormalTok{                        size}\OperatorTok{=}\NormalTok{n\_chromosomes)}
\NormalTok{        p\_lst.append(freq(alleles))}
\NormalTok{        Ht\_lst.append(heterozygosity(alleles))}
    \ControlFlowTok{return}\NormalTok{ p\_lst, Ht\_lst}
\end{Highlighting}
\end{Shaded}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fig, ax }\OperatorTok{=}\NormalTok{ plt.subplots()}
\NormalTok{plt.plot(drift(}\DecValTok{100}\NormalTok{,}\FloatTok{0.5}\NormalTok{))}
\NormalTok{ax.set\_ylabel(}\StringTok{"Frequency of allele $A\_1$, $p$"}\NormalTok{)}
\NormalTok{ax.set\_xlabel(}\StringTok{"Generations"}\NormalTok{)}
\NormalTok{ax.axhline(}\DecValTok{1}\NormalTok{,linestyle}\OperatorTok{=}\StringTok{"{-}{-}"}\NormalTok{, color}\OperatorTok{=}\StringTok{"gray"}\NormalTok{)}
\NormalTok{ax.axhline(}\DecValTok{0}\NormalTok{, linestyle}\OperatorTok{=}\StringTok{"{-}{-}"}\NormalTok{, color}\OperatorTok{=}\StringTok{"gray"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
<matplotlib.lines.Line2D at 0x14bfa2990>
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{pset_1_files/figure-pdf/cell-4-output-2.png}

}

\end{figure}

\hypertarget{problem-2-b}{%
\section{Problem 2 b}\label{problem-2-b}}

\begin{quote}
Simulate 1000 trajectories, keeping a record of the fixation time for
each simulation. Compute the mean time to fixation and compare it to
\(N\). You can run your simulations at 3-5 different values of \(N\) to
establish this dependence better.
\end{quote}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{t\_fixation\_or\_loss }\OperatorTok{=}\NormalTok{ []}
\ControlFlowTok{for}\NormalTok{ N }\KeywordTok{in}\NormalTok{ [}\DecValTok{100}\NormalTok{,}\DecValTok{250}\NormalTok{,}\DecValTok{500}\NormalTok{,}\DecValTok{750}\NormalTok{,}\DecValTok{1000}\NormalTok{]:}
    \ControlFlowTok{for}\NormalTok{ \_ }\KeywordTok{in}\NormalTok{ tqdm(}\BuiltInTok{range}\NormalTok{(}\DecValTok{1000}\NormalTok{)):}
\NormalTok{        p\_lst }\OperatorTok{=}\NormalTok{ drift(N,}\FloatTok{0.5}\NormalTok{)}
\NormalTok{        t\_fixation\_or\_loss.append((N,}\BuiltInTok{len}\NormalTok{(p\_lst)))}
\end{Highlighting}
\end{Shaded}

\hypertarget{problem-2-c}{%
\section{Problem 2 c}\label{problem-2-c}}

\begin{quote}
Plot 10 individual trajectories of heterozygosity vs.~time on a single
graph. In addition, for each time point, compute the average
heterozygosity by averaging over all 1000 trajectories and plot the
average heterozygosity vs time. To better see the exponential decay of
heterozygosity you can plot the log of heterozygosity vs time. Measure
the rate of decay of heterozygosity (the slope on the log-linear plot).
Compare it to N.
\end{quote}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{fig, ax }\OperatorTok{=}\NormalTok{ plt.subplots()}
\ControlFlowTok{for}\NormalTok{ \_ }\KeywordTok{in}\NormalTok{ tqdm(}\BuiltInTok{range}\NormalTok{(}\DecValTok{10}\NormalTok{)):}
\NormalTok{    p\_lst }\OperatorTok{=}\NormalTok{ drift(}\DecValTok{100}\NormalTok{,}\FloatTok{0.5}\NormalTok{)}
\NormalTok{    ax.plot(p\_lst)}
\NormalTok{ax.set\_ylabel(}\StringTok{"Frequency of allele $A\_1$, $p$"}\NormalTok{)}
\NormalTok{ax.set\_xlabel(}\StringTok{"Generations"}\NormalTok{)}
\NormalTok{ax.axhline(}\DecValTok{1}\NormalTok{,linestyle}\OperatorTok{=}\StringTok{"{-}{-}"}\NormalTok{, color}\OperatorTok{=}\StringTok{"gray"}\NormalTok{)}
\NormalTok{ax.axhline(}\DecValTok{0}\NormalTok{, linestyle}\OperatorTok{=}\StringTok{"{-}{-}"}\NormalTok{, color}\OperatorTok{=}\StringTok{"gray"}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  0%|          | 0/10 [00:00<?, ?it/s]100%|██████████| 10/10 [00:00<00:00, 88.44it/s]
\end{verbatim}

\begin{verbatim}
<matplotlib.lines.Line2D at 0x2934a4cd0>
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{pset_1_files/figure-pdf/cell-6-output-3.png}

}

\end{figure}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{lst }\OperatorTok{=}\NormalTok{ []}
\ControlFlowTok{for}\NormalTok{ \_ }\KeywordTok{in}\NormalTok{ tqdm(}\BuiltInTok{range}\NormalTok{(}\DecValTok{1000}\NormalTok{)):}
\NormalTok{    p\_lst, Ht\_lst }\OperatorTok{=}\NormalTok{ drift(}\DecValTok{100}\NormalTok{,}\FloatTok{0.5}\NormalTok{)}
\NormalTok{    lst.append(Ht\_lst)}

\NormalTok{lens }\OperatorTok{=}\NormalTok{ np.array([}\BuiltInTok{len}\NormalTok{(i) }\ControlFlowTok{for}\NormalTok{ i }\KeywordTok{in}\NormalTok{ lst])}

\CommentTok{\# Mask of valid places in each row}
\NormalTok{mask }\OperatorTok{=}\NormalTok{ np.arange(lens.}\BuiltInTok{max}\NormalTok{()) }\OperatorTok{\textless{}}\NormalTok{ lens[:,}\VariableTok{None}\NormalTok{]}

\NormalTok{out }\OperatorTok{=}\NormalTok{ np.zeros(mask.shape)}
\NormalTok{out[mask] }\OperatorTok{=}\NormalTok{ np.concatenate(lst)}
\NormalTok{mean\_heterozygosity }\OperatorTok{=}\NormalTok{ out.mean(axis}\OperatorTok{=}\DecValTok{0}\NormalTok{)}

\NormalTok{log\_mean\_heterozygosity }\OperatorTok{=}\NormalTok{ np.log(mean\_heterozygosity)}

\NormalTok{fig, ax }\OperatorTok{=}\NormalTok{ plt.subplots()}
\NormalTok{ax.plot(mean\_heterozygosity)}
\NormalTok{ax.set\_ylabel(}\StringTok{"Heterozygosity $H\_t$"}\NormalTok{)}
\NormalTok{ax.set\_xlabel(}\StringTok{"Generations"}\NormalTok{)}

\NormalTok{fig, ax }\OperatorTok{=}\NormalTok{ plt.subplots()}
\NormalTok{ax.plot(log\_mean\_heterozygosity)}
\NormalTok{ax.set\_ylabel(}\VerbatimStringTok{r"Log heterozygosity $\textbackslash{}log\textbackslash{}left[H\_t\textbackslash{}right]$"}\NormalTok{)}
\NormalTok{ax.set\_xlabel(}\StringTok{"Generations"}\NormalTok{)}


\NormalTok{df }\OperatorTok{=}\NormalTok{ pd.DataFrame()}
\NormalTok{df[}\StringTok{"Log"}\NormalTok{] }\OperatorTok{=}\NormalTok{ log\_mean\_heterozygosity}
\NormalTok{df[}\StringTok{"Generations"}\NormalTok{] }\OperatorTok{=}\NormalTok{ np.arange(}\DecValTok{1}\NormalTok{,}\BuiltInTok{len}\NormalTok{(mean\_heterozygosity)}\OperatorTok{+}\DecValTok{1}\NormalTok{)}
\NormalTok{model }\OperatorTok{=}\NormalTok{ smf.ols(}\StringTok{"Log \textasciitilde{} Generations"}\NormalTok{, data }\OperatorTok{=}\NormalTok{ df[:}\OperatorTok{{-}}\DecValTok{250}\NormalTok{])}
\NormalTok{results }\OperatorTok{=}\NormalTok{ model.fit()}

\NormalTok{ax.plot(results.predict(df[}\StringTok{"Generations"}\NormalTok{]))}
\NormalTok{ax.text(}\DecValTok{1000}\NormalTok{,}\OperatorTok{{-}}\DecValTok{2}\NormalTok{,}\StringTok{"$R\^{}2$=}\SpecialCharTok{\{:.2\}}\StringTok{"}\NormalTok{.}\BuiltInTok{format}\NormalTok{(results.rsquared))}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
  0%|          | 0/1000 [00:00<?, ?it/s]100%|██████████| 1000/1000 [00:12<00:00, 79.04it/s]
/var/folders/_c/3843qt453ds6klt3y95rpjcw0000gn/T/ipykernel_40568/3978446525.py:15: RuntimeWarning: divide by zero encountered in log
  log_mean_heterozygosity = np.log(mean_heterozygosity)
\end{verbatim}

\begin{verbatim}
Text(1000, -2, '$R^2$=0.99')
\end{verbatim}

\begin{figure}[H]

{\centering \includegraphics{pset_1_files/figure-pdf/cell-7-output-3.png}

}

\end{figure}

\begin{figure}[H]

{\centering \includegraphics{pset_1_files/figure-pdf/cell-7-output-4.png}

}

\end{figure}

The log heterozygosity is largely linear with negative gradient and
implies an exponential decay of the heterozygosity.

\[H_{t}=H_0\left(1-\frac{1}{2N}\right)^{t}\approx H_0 \exp\left[{-\frac{t}{2N}}\right]\]

\[\log H_{t}=\log H_0 -\frac{t}{2N}\]

\hypertarget{problem-2-d}{%
\section{Problem 2 d}\label{problem-2-d}}

\begin{quote}
Consider an effect of changing population size on the time to fixation.
Run simulations first at some \(N=N_1\) and then change \(N\) to
\(N=N_2\) for \(t_2\) number of generations then back to \(N_1\). Try
\(N_2\ll N_1\), e.g.~\(N2=N1/10\). Make 1000 runs and compute the mean
time to fixation. Try several \(N_1\) and \(N_2\). How does decrease in
the population size change the mean time to fixation? Compute
\(N_\mathrm{eff}\) for each parameters (\(N_1\), \(N_2\), \(t_2\)), and
correlated mean time to fixation with \(N_1\), \(N_2\), and
\(N_\mathrm{eff}\). Interpret your findings.
\end{quote}

\hypertarget{problem-2-e}{%
\section{Problem 2 e}\label{problem-2-e}}

\begin{quote}
{[}Extra credit: +2 points{]} Plot the distribution of the fixation time
for each N. How broad is the distribution? Think of various ways to
quantify this. Study the mean time to fixation as a function of the
number K of alleles (types of individuals), starting with 1/K initial
frequency.
\end{quote}



\end{document}
